modules/javafx.graphics/src/main/native-glass/gtk/glass_window_ime.cpp

```diff
@@ -23,266 +23,157 @@
-    static size_t buf_len = 12;
-    static char *buffer = NULL;
+    gtk_im_context_get_preedit_string(im_context, &preedit_text, &attrList, &cursor_pos);
+    ctx->updateCaretPos();
 
-    if (buffer == NULL) {
-        buffer = (char*)malloc(buf_len * sizeof (char));
-    }
+    jstring jstr = mainEnv->NewStringUTF(preedit_text);
+    EXCEPTION_OCCURED(mainEnv);
 
-    if (buffer == NULL) {
-        // dont process the key event
-        fprintf(stderr, "malloc failed in im_filter_keypress\n");
-        return false;
-    }
+    jsize slen = mainEnv->GetStringLength(jstr);
 
-    KeySym keysym;
-    Status status;
-    XKeyPressedEvent xevent = convert_event(event);
-    if (XFilterEvent((XEvent*) & xevent, GDK_WINDOW_XID(gdk_window))) {
-        return TRUE;
-    }
-
-    if (event->type == GDK_KEY_RELEASE) {
-        process_key(event);
-        return TRUE;
-    }
-
-    int len = Xutf8LookupString(xim.ic, &xevent, buffer, buf_len - 1, &keysym, &status);
-    if (status == XBufferOverflow) {
-        buf_len = len + 1;
-        char *tmpBuffer = (char*)realloc(buffer, buf_len * sizeof (char));
-        if (tmpBuffer == NULL) {
-            SAFE_FREE(buffer);
-            // dont process the key event
-            fprintf(stderr, "realloc failed in im_filter_keypress\n");
-            return false;
-        }
-        buffer = tmpBuffer;
-        len = Xutf8LookupString(xim.ic, &xevent, buffer, buf_len - 1,
-                &keysym, &status);
-    }
-    switch (status) {
-        case XLookupKeySym:
-        case XLookupBoth:
-            if (xevent.keycode) {
-                //process it as a normal key
-                process_key(event);
-                break;
-            }
-            // fall-through
-        case XLookupChars:
-            buffer[len] = 0;
-            jstring str = mainEnv->NewStringUTF(buffer);
-            EXCEPTION_OCCURED(mainEnv);
-            jsize slen = mainEnv->GetStringLength(str);
-            mainEnv->CallVoidMethod(jview,
-                    jViewNotifyInputMethod,
-                    str,
-                    NULL, NULL, NULL,
-                    slen,
-                    slen,
-                    0);
-            LOG_EXCEPTION(mainEnv)
+    PangoAttrIterator *iter = pango_attr_list_get_iterator(attrList);
+    PangoAttribute *pangoAttr;
 
+    jbyte attr = com_sun_glass_ui_View_IME_ATTR_INPUT;
+    do {
+        if (pangoAttr = pango_attr_iterator_get(iter, PANGO_ATTR_BACKGROUND)) {
+             attr = com_sun_glass_ui_View_IME_ATTR_TARGET_NOTCONVERTED;
+             break;
+        } else if ((pangoAttr = pango_attr_iterator_get(iter, PANGO_ATTR_UNDERLINE))
+                && (((PangoAttrInt *)pangoAttr)->value == PANGO_UNDERLINE_SINGLE)) {
+            attr = com_sun_glass_ui_View_IME_ATTR_CONVERTED;
             break;
-    }
+        }
+    } while (pango_attr_iterator_next(iter));
+
+    pango_attr_iterator_destroy (iter);
```

Comment:
According to the docs you need to release the attribute list using pango_attr_list_unref().

Reply:
You're right. Fixed.